### ISSUE TITLE
introduce setters/getters for default parameters of each object

### ISSUE THREAD

**Issue by gonuke** [svalinn/parastell#60]

**Comment by connoramoreno** [comment]
I'm trying to decide how to implement setters and getters for each object.

In my current implementation, each object has corresponding setter and getter functions that are member functions of the `Stellarator` class in `parastell.py`. Default parameters are still needed to supplement the YAML configuration of each object in its respective script, however (the `oo_version` branch of ParaStell uses the default dictionaries defined in `utils.py` to supplement YAML configurations). With this methodology, those setter/getter function are not accessible from each object's script. I can duplicate corresponding setter and getter functions in each script, but this doesn't feel very streamlined.

Alternatively, I can define a non-member "setter" function in each object's script that takes the input dictionary as an argument and supplements any missing keys with default values. These "setter" functions could then be called in each script to supplement corresponding YAML configurations, as well as be called by `parastell.py` for all of its default parameter needs.

Let me know if I need to clarify any of this; also open to alternate suggestions.

**Comment by gonuke** [comment]
I was thinking something different...   Take the `InvesselBuild` class and let's assume that the `plasma_mat_tag` will be `plasma` by default.

The `__init__` function would no longer take an argument for `_plasma_mat_tag`.  Instead, it would always set `self._plasma_mat_tag` to be `plasma`.  (Note, I've prepended an underscore for reasons that will be apparent below.

There would also be the following functions:
```
@property
def plasma_mat_tag(self):
    return self._plasma_mat_tag

@plasma_mat_tag.setter
def plasma_mat_tag(self, mat_tag):
    self._plasma_mat_tag = mat_tag
```

We could do it without these functions and just let folks set the member variable directly (they aren't hidden if we leave the underscore off) but if we do use functions, we can theoretically add some kind of checks, etc.  Maybe we don't need that...?

reference: https://realpython.com/python-getter-setter/

**Comment by connoramoreno** [comment]
I see, I misunderstood what you were envisioning. That makes sense to me! Then if users wanted to set parameters themselves, say `radial_build`,  they could use
```python
import parastell.invessel_build as ivb

radial_build = ...

invessel_build = ivb.InVesselBuild()
invessel_build.radial_build = radial_build
```

or if using `parastell.py`,

```python
import parastell.parastell as ps

radial_build = ...

stellarator = ps.Stellarator()
stellarator.populate_invessel_build()
stellarator.invessel_build.radial_build = radial_build
stellarator.construct_invessel_build()
...
```
Note that I included an additional method in the `Stellarator` class that would instantiate `InVesselBuild` since users would need to be able to modify parameters before any actual construction.

Alternatively, each object (`InVesselBuild`, etc.) could be instantiated by default to get rid of need for functions that instantiate them.

**Comment by connoramoreno** [comment]
`radial_build` is a bad example since that might not have a default value but hopefully my thoughts are still clear

**Comment by gonuke** [comment]
There may still be value in passing some things to the constructor so that the non-default things can be setup all at once.

**Comment by connoramoreno** [comment]
I'm having the essential inputs passed to the constructor as arguments, but still using setters/getters for all input parameters (arguments and default parameters).

I'm liking setters as a way of checking and modifying input parameters.

**Comment by connoramoreno** [comment]
The only potential issue I'm anticipating is setting these "default" parameters via YAML, since there's no way of interacting with the class objects directly. I suppose in the step that extracts the YAML data, I could have a series of `if` statements that conditionally set the default parameters if present in the YAML configuration.

**Comment by gonuke** [comment]
Hmmm....

So what about something like this:

``` pseudocode
class InvesselBuild(object):

   def __init__():
        # sets default values for all things requiring defaults

    # various setters/getters and/or expose variables directly to user

    def init_from_dict(self, parameter_dict):
        for key, value in parameter_dict.items():
            if key in allowed_keys:
                self.__setattr__(key) = value
            else:
                raise ValueError(f'"{key}" is not a supported parameter of InvesselBuild')
```



```

**Comment by connoramoreno** [comment]
That's a great idea! Thanks!

**Comment by connoramoreno** [comment]
Closed by #85.
